pkg/streamingpromql/operators/functions/histograms.go

```diff
@@ -0,0 +1,410 @@
+	if len(fPoints) > 0 {
+		upperBound, err := strconv.ParseFloat(le, 64)
+		if err != nil {
+			// The le label was invalid. Record it:
+			h.annotations.Add(annotations.NewBadBucketLabelWarning(
+				h.innerSeriesMetricNames.GetMetricNameForSeries(h.currentInnerSeriesIndex),
+				le,
+				h.inner.ExpressionPosition(),
+			))
+		} else {
+			if g.pointBuckets == nil {
+				g.pointBuckets, err = pointBucketPool.Get(h.timeRange.StepCount, h.memoryConsumptionTracker)
+				if err != nil {
+					return err
+				}
+				g.pointBuckets = g.pointBuckets[:h.timeRange.StepCount]
+			}
+			for _, f := range fPoints {
+				pointIdx := h.timeRange.PointIndex(f.T)
+				g.pointBuckets[pointIdx] = append(
+					g.pointBuckets[pointIdx],
+					bucket{
+						upperBound: upperBound,
+						count:      f.F,
+					},
+				)
+			}
+		}
+	}
+	g.remainingSeriesCount--
+	return nil
+}
+
+// saveNativeHistogramsToGroup stores the given native histograms onto the given group.
+// There should only ever be one native histogram per a series or series group. In general,
+// they are per-series, but we handle them as parts of groups because they could hypothetically
+// have the same series name as a classic histogram.
+func (h *HistogramFunctionOverInstantVector) saveNativeHistogramsToGroup(hPoints []promql.HPoint, g *bucketGroup) {
+	if len(hPoints) > 0 {
+		// We should only ever see one set of native histograms per group.
+		if g.nativeHistograms != nil {
+			panic("We should never see more than one native histogram per group")
+		}
+		g.nativeHistograms = hPoints
+	}
+	g.remainingSeriesCount--
+}
+
+func (h *HistogramFunctionOverInstantVector) computeOutputSeriesForGroup(g *bucketGroup) (types.InstantVectorSeriesData, error) {
+	floatPoints, err := types.FPointSlicePool.Get(h.timeRange.StepCount, h.memoryConsumptionTracker)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	histogramIndex := 0
+	var currentHistogram *histogram.FloatHistogram
+	var currentHistogramPointIdx int64
+
+	for pointIdx := range h.timeRange.StepCount {
+		ph := h.phValues.Samples[pointIdx].F
+		if math.IsNaN(ph) || ph < 0 || ph > 1 {
+			// Even when ph is invalid we still return a series as BucketQuantile will return +/-Inf.
+			// So don't skip/continue the series, just emit a warning.
+			h.annotations.Add(annotations.NewInvalidQuantileWarning(ph, h.phArg.ExpressionPosition()))
+		}
+
+		// Get the HPoint if it exists at this step
+		if g.nativeHistograms != nil && histogramIndex < len(g.nativeHistograms) {
+			nextHPoint := g.nativeHistograms[histogramIndex]
+			if h.timeRange.PointIndex(nextHPoint.T) == int64(pointIdx) {
+				currentHistogram = nextHPoint.H
+				currentHistogramPointIdx = h.timeRange.PointIndex(nextHPoint.T)
+				histogramIndex++
+			}
+		}
+
+		if g.pointBuckets != nil && len(g.pointBuckets[pointIdx]) > 0 && currentHistogram != nil && currentHistogramPointIdx == int64(pointIdx) {
```

Comment:
Couldn't we move the declaration of `currentHistogram` inside the body of the `for` and then simplify this to:

```suggestion
		if g.pointBuckets != nil && len(g.pointBuckets[pointIdx]) > 0 && currentHistogram != nil {
```

Then `currentHistogram` will only ever be non-nil if it is for the current time step, and `currentHistogramPointIdx` can be removed.